File: python/day4.py
lines = open("../inputs/day4.txt").read().strip().split("\n")

part1 = 0
part2 = 0

for line in lines:
    a, b = line.split(",")
    start1, end1 = map(int, a.split("-"))
    start2, end2 = map(int, b.split("-"))

    if (start1 <= start2 and end1 >= end2) or (start2 <= start1 and end2 >= end1):
        part1 += 1

    if max(start1, start2) <= min(end1, end2):
        part2 += 1

if __name__ == "__main__":
    print(part1)
    print(part2)

File: javascript/day4.js
import fs from 'fs'

const rangePairs = fs
  .readFileSync('../inputs/day4.txt')
  .toString('utf-8')
  .split('\n')
  .filter((l) => l)

const parseRangePair = (rangePair) => {
  const match = rangePair
    .match(/(\d+)-(\d+),(\d+)-(\d+)/)
    .slice(1)
    .map(Number)
  return match
}

const checkContaining = ([start1, end1, start2, end2]) =>
  (start1 <= start2 && end1 >= end2) || (start2 <= start1 && end2 >= end1)

const checkOverlaping = ([start1, end1, start2, end2]) =>
  Math.max(start1, start2) <= Math.min(end1, end2)

const part1 = rangePairs.map(parseRangePair).filter(checkContaining).length
const part2 = rangePairs.map(parseRangePair).filter(checkOverlaping).length

console.log(part1)
console.log(part2)
